### PR TITLE
Add support for properly tagging images and pushing them, based on the v...

### PR DESCRIPTION
**Pull request by josephschorr**
...arious levels of configuration for jobs

### CODE REVIEW

File: _example/main.go [hunk 1]
```diff
@@ -21,18 +20,21 @@ blocks:
...
-	options.SubDirectory = "testmultibuild-master"
-
-	err := buildcontroller.InvokeBuild(testConfig, options)
+	err := buildcontroller.InvokeBuild(testConfig, buildcontroller.InvokeBuildOptions{
+		BuildPackage: buildcontroller.BuildPackageOptions{
+			URL:          "https://github.com/josephschorr/testmultibuild/archive/master.zip",
+			SubDirectory: "testmultibuild-master",
+		},
+		ProjectName: "exampleproject",
+		Registry:    "quay.io/sylphon",
```
Comment (colegleason) [inline _example/main.go:36]: Are we sure we don't want these set up in the config? We have the credentials there, so it seems odd to list the registry here.

File: controller.go [hunk 2]
```diff
@@ -33,32 +48,65 @@ func InvokeBuild(yamlConfig string, buildPackage BuildPackageOptions) error {
...
+	username := options.DefaultPushCredentials.Username
+	password := options.DefaultPushCredentials.Password
+
+	if job.PushInfo.Credentials.Username != "" {
+		username = job.PushInfo.Credentials.Username
+	}
+
+	if job.PushInfo.Credentials.Password != "" {
+		password = job.PushInfo.Credentials.Password
+	}
```
Comment (colegleason) [inline controller.go:87]: Does it make sense to do these in separate if statements? What if one is defined but the other is not?

File: _example/main.go (hunk 1 above)
Comment (josephschorr) [inline _example/main.go:36]: We should. We need it in the build graph config at both levels

File: controller.go (hunk 2 above)
Comment (josephschorr) [inline controller.go:87]: I can put them together, if it makes more sense

File: _example/main.go (hunk 1 above)
Comment (colegleason) [inline _example/main.go:36]: You mean in the "blocks" list and in the top-level as a default?

File: controller.go (hunk 2 above)
Comment (colegleason) [inline controller.go:87]: It probably doesn't matter that much, but it doesn't really make sense to have a default password but be overriding the username each time.

File: _example/main.go (hunk 1 above)
Comment (colegleason) [inline _example/main.go:36]: And if that is the case, would you like graph-builder to fill in the registry at the job level or will the controller also parse the YAML to find top-level defaults?

File: controller.go (hunk 2 above)
Comment (josephschorr) [inline controller.go:87]: Sounds good

File: _example/main.go (hunk 1 above)
Comment (josephschorr) [inline _example/main.go:36]: The controller can handle it. The graph builder should present the parsed YAML directly (like it does now)

**Comment by josephschorr** [comment]
@rafecolton @colegleason Any other fixes before this gets merged?

**Comment by colegleason** [comment]
Nope!

**Comment by rafecolton** [comment]
LGTM

**Pull request merged**